### ISSUE TITLE
VolumeSnapshotContent update fails with "object has been modified"

### ISSUE THREAD

**Issue by ggriffiths** [kubernetes-csi/external-snapshotter#525]
Seeing this consistently on snapshotting e2e tests in the snapshot controller pod:

```
~ curl -s https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-csi_external-snapshotter/519/pull-kubernetes-csi-external-snapshotter-1-21-on-kubernetes-1-21/1396873247164731392/artifacts/kube-system/snapshot-controller-58f54d8f6c-9gt8c/snapshot-controller.log |  grep 'the object has been modified; please apply your' | wc -l
      56
```

There seems to be multiple root causes for this error. We'll be tracking them here:
- [x] [updateContentErrorStatusWithEvent](https://github.com/kubernetes-csi/external-snapshotter/blob/679e71e89d4644ae4b50fa71874b4d9a77bb9e5a/pkg/sidecar-controller/snapshot_controller.go#L140) - ~~if we perform a Get before Update here, the errors seem to drop by half.~~ Will not do this initially as it will lower performance.
- [x] [functions not using or returning updatedContent after update calls](https://github.com/kubernetes-csi/external-snapshotter/blob/8751eb0ca5b33a8a3c2589d506b48cc26684e0cd/pkg/sidecar-controller/snapshot_controller.go#L599) - various functions do not propagate their `updatedContent` back up the stack to be cached properly (see below for explanation). https://github.com/kubernetes-csi/external-snapshotter/pull/527
- [ ] JSONPatch can also be used to lower the error count - https://github.com/kubernetes-csi/external-snapshotter/pull/526

**Comment by jsafrane** [comment]
Debugging the case when `CreateSnapshot` returns a final error and this error is not saved to VolumeSnapshotContent because of "object has been modified". This happens when e.g. related VolumeSnapshotClass has an invalid parameter and the CSI driver checks for it, e.g. Azure Disk. (Not reproducible with AWS EBS, GCP PD, csi-hostpath, they don't check parameters...)

external-snapshotter sidecar logs:

```I0527 13:05:05.027686       1 snapshot_controller.go:56] synchronizing VolumeSnapshotContent[snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4]
I0527 13:05:05.027695       1 snapshot_controller.go:540] Check if VolumeSnapshotContent[snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4] should be deleted.
I0527 13:05:05.027703       1 snapshot_controller.go:75] syncContent: Call CreateSnapshot for content snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4
[snip]
I0527 13:05:05.030255       1 connection.go:186] GRPC error: rpc error: code = Unknown desc = failed create snapshot: exit status 1: tar: can't change directory to '/csi-data-dir/a0af7d68-beeb-11eb-b974-0a580a830067': No such file or directory
I0527 13:05:05.030273       1 snapshot_controller.go:301] createSnapshotWrapper: CreateSnapshot for content snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4 returned error: rpc error: code = Unknown desc = failed create snapshot: exit status 1: tar: can't change directory to '/csi-data-dir/a0af7d68-beeb-11eb-b974-0a580a830067': No such file or directory
I0527 13:05:05.216306       1 request.go:591] Throttling request took 185.901962ms, request: PUT:https://172.30.0.1:443/apis/snapshot.storage.k8s.io/v1/volumesnapshotcontents/snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4
I0527 13:05:05.230174       1 snapshot_controller_base.go:143] enqueued "snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4" for sync
I0527 13:05:05.231459       1 snapshot_controller.go:614] Removed VolumeSnapshotBeingCreated annotation from volume snapshot content snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4
I0527 13:05:05.231486       1 util.go:297] storeObjectUpdate updating content "snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4" with version 117229
I0527 13:05:05.231507       1 snapshot_controller.go:141] updateContentStatusWithEvent[snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4]
I0527 13:05:05.416339       1 request.go:591] Throttling request took 184.681792ms, request: PUT:https://172.30.0.1:443/apis/snapshot.storage.k8s.io/v1/volumesnapshotcontents/snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4/status
I0527 13:05:05.426486       1 snapshot_controller.go:165] updating VolumeSnapshotContent[snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4] error status failed Operation cannot be fulfilled on volumesnapshotcontents.snapshot.storage.k8s.io "snapcontent-3e1d8b31-98d2-4b7a-847f-ef99d40fd3e4": the object has been modified; please apply your changes to the latest version and try again
```

The external-snapshotter removes `VolumeSnapshotBeingCreated` annotation and then it tries to update `VolumeSnapshotContent` again with error message. This second update fails, because the sidecar does not use `VolumeSnapshotContent` returned when removing the annotation.

In other words, `removeAnnVolumeSnapshotBeingCreated` should return `updatedContent` from `VolumeSnapshotContents().Update()`:

https://github.com/kubernetes-csi/external-snapshotter/blob/8751eb0ca5b33a8a3c2589d506b48cc26684e0cd/pkg/sidecar-controller/snapshot_controller.go#L607

And this `*VolumeSnapshotContent` should be returned all the way up the stack to be passed into `updateContentErrorStatusWithEvent` here:

https://github.com/kubernetes-csi/external-snapshotter/blob/8751eb0ca5b33a8a3c2589d506b48cc26684e0cd/pkg/sidecar-controller/snapshot_controller.go#L121

Beware of returning `nil` along the path... IMO all `*VolumeSnapshotContent` returns should return *some* valid object, either the one that was saved to the API server last, or the original one. You might also use `ctrl.contentStore` to get the latest version, I did not explore this path.

cc @huffmanca

**Comment by ggriffiths** [comment]
Still planning to further address this issue in the next few weeks. The next step will be to fix up #526.

**Comment by ggriffiths** [comment]
/remove-lifecycle stale 
working on this issue now.
